### PR TITLE
Testcast to demonstrate issue with #1188

### PR DESCRIPTION
**Pull request by mfazekas**
Testcase demonstrating issue with #1188 as well as current find_in_unresolved tree.

Since #1188 makes find_in_unresolved a 1 level deep search only, the ```d``` in example will be found by the ```try_activate``` which will just look in the latest version of gems and will not care if it conflicts with an unresolved dependency.

The issue with with current find_in_unresolved is that we should prefer latest version so we should be seeing ```a-1 b-2 c-2 d-2``` and not ```a-1 b-2 c-1 d-1```

### CODE REVIEW

**Comment by tenderlove** [comment]
@mfazekas awesome,  thank you so much.  I *really* appreciate the help with this stuff.  Given [your comment here](https://github.com/rubygems/rubygems/pull/1188#issuecomment-82419621), do you still think we should revert #1188?  `master` is significantly faster (even though it doesn't behave the same way), and I think we can even rm most of the code I wrote and still have tests pass.  I think you have a better understanding of the impacts of #1188, so I will go with your recommendation.

FWIW, I think we can revert #1188 and still get some speed improvements through caching.  I was experimenting with some caches and was able to cut your test time in half (though "half time" was still  too slow IMO).

**Comment by mfazekas** [comment]
@tenderlove, yep because the current traverse code is broken soo many ways it's probably not too much usefull. So we can probably get away with replacing with some other broken implementation that's not exponentiental at least. I don't like the current #1188 is because it does an ```nop``` statement  with```... unless expr``` where expr is true and thats not a nice or easy to read idiom.

So the original code tries to enumerate all possible path in the dependency graph and uses muliple reverse so it selects the latest version of all those non conflicting paths. But it's slow, and doesn't select the latest and can select conflict. Ideally we should do some greedy depth first search and just return the first path we find good. I'll try to collect all my testcases regarding this and write something over the weekend.

Or we could just make traverse a noop, or just one level as #1188 did, and there is a good chance we're able to get away with that too. 

It's hard to know which gem combination people have and which is effected by the change.

**Comment by evanphx** [comment]
Since we want to overhaul (in multiple ways) this code, maybe we should focus on building up the tests so that we can properly evaluate different approaches. I've always worried that we don't have enough test coverage over resolution.

**Comment by tenderlove** [comment]
> Or we could just make traverse a noop, or just one level as #1188 did, and there is a good chance we're able to get away with that too.

ya, that's what I was thinking when I said "I think we can even rm most of the code I wrote and still have tests pass"

> Since we want to overhaul (in multiple ways) this code, maybe we should focus on building up the tests so that we can properly evaluate different approaches. I've always worried that we don't have enough test coverage over resolution.

I think we need to get a clear understanding of *what we want the code to do*.  @mfazekas demoed in #1169 that the resolution before #1188 is just broken.  This PR demonstrates that it's still broken, just in a different way. #1169 demonstrates that before #1188 it was not just broken, but *really* slow.  After #1188 it's still broken but not slow.

Repo  | Slow | Broken
------------- | ------|-------
Before #1188  | Yes | Yes
After #1188  | No | Yes (but differently)

From a high level perspective, I don't believe it's possible to correctly resolve a dependency tree at runtime like this code is trying to do (can someone prove me wrong?).  IMO we should just make a "best effort" and if you've got a complex setup suggest people use a Gemfile.  I also noted that the new resolver PR #1189 doesn't seem to touch this code, so I'm starting to suspect that RG actually contains two resolvers.

TL;DR: I think we should sacrifice correctness for speed.

**Comment by mfazekas** [comment]
I aggree that we cannot make config resolution 100% as it might depend on what they actually require at runtime.

I'd say requirements:
1. should prefer gems in direct dependencies
2. should prefer gems in indirect dependenices to gems outside the dependency tree
3. should make reasonable effort to avoid conflicts
4. should prefer latest versions
5. should be performant

And we're talking about commands like ```rails new``` and ```heroku``` where requiring bundler might not be practical - and anyway bundler is not shipping with rails rubygems is so it has to cover most of cases.

> 1 should prefer gems in direct dependencies

Example: if i declare that my gem depends on 'rake' and i require 'rake', we need to find it there even if my system has an 'antirake' gem containing a 'rake.rb' too.

> 2 should prefer gems in indirect dependenices to gems outside the dependency tree

Example: if i declare that my gem depends on 'rails ~> 4.2.1' and i ```require 'activerecord'``` should find in rails '~ 4.2.1' compatible activerecords. 

> 3 should make reasonable effort to avoid conflicts

If i declare rails '~> 4.2.1' it's reasonable to expect that ```require 'activerecord'```  will be found in an activerecord gem compatible with those rails versions.

((and for deeper level example would be rails -> railities -> rake. So if i declared rails as a dependency and require 'rake' we should find it in latest rake version compatible with the rails))

I've added a series of tests that evaluated both for original and #1188.
https://github.com/mfazekas/rubygems/commit/223e20f859d01b8974f5a568721c8a509843b784
In comment you can see the correct result, the result returned by #1188 and the result returned by original.

```
{
  #correct: %w(a-1 b-2 c-2 d-2),     
  tenderlove: %w(a-1 d-2),
  original: %w(a-1 b-2 c-1 d-1)
}
```

**Comment by mfazekas** [comment]
Added 2 more variations at:

1.) original_fixed: just fixes the issue with the original (it uses correct number of reverse to search for the latest), and it correctly skips conflicting gems (original used next incorrectly on conflict). This passes all tests, but slow.
2.) mfazekekas: it memoizes if we already seen a gem and we dont revisit again, this makes the alg non exponentiental - i think it's O(n^2) becuase of array append, which could be improved with linked lists. We also do greedy and start with latest versions, rather than collecting a lot then selecting the latest. This passess all tests and fast. (I'll try to came up a testcase that the original_fixed passes but this one fails)

https://github.com/mfazekas/rubygems/commit/31cb140068f9a9a865c8ceadab6be6a8a30458ad

**Comment by tenderlove** [comment]
@mfazekas sounds great.  If mfazekas@31cb140 passes all the tests, can you clean up the patches and send a PR?  I agree with the requirements you laid out, and the resolution changes you made in the tests for your patch make more sense to me. :+1:

**Comment by mfazekas** [comment]
Unfortunately there was other issue with original impl [has_conflicts?](https://github.com/rubygems/rubygems/blob/c2aebd9b3af82c969711032f1a36ac35d55337cb/lib/rubygems/specification.rb#L966), is only if the item dependecies conflicts with loaded gem, but not if he item itself is confilicts with any of the gem to be loaded (eg trail). 

I've added [that check](https://github.com/mfazekas/rubygems/commit/32e314173990ba4756b7354fe51053e5e11d4495#diff-05477d38d75c98664b0e7dec6eb73b45R1029) in my impl but it means in same case we'll see some perf degradation. In the original perf case we're still much much faster (<0.1 sec vs 20+sec) but with a [new perf test case](https://github.com/mfazekas/rubygems/commit/32e314173990ba4756b7354fe51053e5e11d4495#diff-6d1d09cbeb26123575a4efe6dc0ddab2R376) where we have a lot of gems have one version and they have a chain. Original was 7.75 sec while new is 8.08 sec so not too terrible but a few percent. This is for 2000 includes, and 100 gems.

**Comment by mfazekas** [comment]
Added a PR in #1202

**Comment by mfazekas** [comment]
see #1202
